### PR TITLE
Refactor join step

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Processors/QueryPipeline.cpp [hunk 1]
```diff
@@ -255,6 +262,95 @@ QueryPipeline QueryPipeline::unitePipelines(
...
+    }
+
+    ///                                     (left)   ──────> Joining ─> (joined)
+    ///                                     (left)   ─┐┌───>
+    ///                                   (totals)   ─┼┼─┐
+    ///                                               └┼─┼─> Joining ─> (joined)
+    ///                                                │┌┼─>
+    /// (right)  ─>                                  ──┘││
+    /// (right)  ─> Resize ─> AddingJoined ─> Resize ───┘└─> Joining ─> (totals)
+    /// (totals) ───────────>                        ──────>
```
Comment (akuzm) [inline src/Processors/QueryPipeline.cpp:297]: Nice diagram. Two entries for (right) and Joining mean parallel execution, right? I didn't realize right away where the splits/merges are, maybe it makes sense to picture them with `┤` and the like.

File: src/Processors/Transforms/JoiningTransform.h [hunk 2]
```diff
@@ -33,7 +75,33 @@ class JoiningTransform : public ISimpleTransform
...
+    BlockInputStreamPtr non_joined_stream;
+    size_t max_block_size;
+
     Block readExecute(Chunk & chunk);
 };
 
+/// Fills Join with block from right table.
+/// Has single input and single output port.
+/// Output port has empty header. It is closed when al data is inserted in join.
+class AddingJoinedTransform : public IProcessor
```
Comment (akuzm) [inline src/Processors/Transforms/JoiningTransform.h:88]: The name sounds cryptic, maybe something like `PrepareRightJoinSideTransform`?

File: src/Processors/Transforms/JoiningTransform.h [hunk 3]
```diff
@@ -8,21 +8,63 @@ namespace DB
...
 
-class JoiningTransform : public ISimpleTransform
+class IBlockInputStream;
+using BlockInputStreamPtr = std::shared_ptr<IBlockInputStream>;
+
+/// Join rows to chunk form left table.
+/// This transform usually has two input ports and one output.
+/// First input is for data from left table.
+/// Second input has empty header and is connected with AddingJoined.
+/// We can process left table only when Join is filled. Second input is used to signal that AddingJoined is finished.
```
Comment (akuzm) [inline src/Processors/Transforms/JoiningTransform.h:18]: How does it handle switching to merge joins? For hash joins we first have to process the entire right side, but for merge joins, we process left and right sides simultaneously.

File: src/Processors/QueryPipeline.cpp (hunk 1 above)
Comment (KochetovNicolai) [reply src/Processors/QueryPipeline.cpp:297]: > Two entries for (right) and Joining mean parallel execution, right?

Yes.

> maybe it makes sense to picture them with ┤ and the like

Good idea!

File: src/Processors/Transforms/JoiningTransform.h (hunk 3 above)
Comment (KochetovNicolai) [reply src/Processors/Transforms/JoiningTransform.h:18]: For now, we have only "partial merge join". It also reads right table first.
Looks like now we can implement honest merge join

File: src/Processors/QueryPipeline.cpp (hunk 1 above)
Comment (KochetovNicolai) [reply src/Processors/QueryPipeline.cpp:297]: I've spend like 20 minutes trying to make it better, but unicode seem to be not so powerful...

File: src/Processors/QueryPlan/JoinStep.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    , max_block_size(max_block_size_)
+{
+    input_streams = {left_stream_, right_stream_};
+    output_stream = DataStream
+    {
+        .header = JoiningTransform::transformHeader(left_stream_.header, join),
+    };
+}
+
+QueryPipelinePtr JoinStep::updatePipeline(QueryPipelines pipelines, const BuildQueryPipelineSettings &)
```
Comment (vdimir) [inline src/Processors/QueryPlan/JoinStep.cpp:29]: What difference between `updatePipeline` and `transformPipeline`? Why JoinStep uses update and FilledJoinStep uses transform?

File: src/Processors/Transforms/JoiningTransform.h [hunk 5]
```diff
@@ -8,21 +8,63 @@ namespace DB
...
 using JoinPtr = std::shared_ptr<IJoin>;
 
-class JoiningTransform : public ISimpleTransform
+class IBlockInputStream;
+using BlockInputStreamPtr = std::shared_ptr<IBlockInputStream>;
+
+/// Join rows to chunk form left table.
+/// This transform usually has two input ports and one output.
+/// First input is for data from left table.
+/// Second input has empty header and is connected with AddingJoined.
```
Comment (vdimir) [inline src/Processors/Transforms/JoiningTransform.h:17]: AddingJoined looks like class name, but I have not found some entity with such name. What does it refer to?

File: src/Interpreters/ExpressionAnalyzer.cpp [hunk 6]
```diff
@@ -802,67 +807,77 @@ JoinPtr SelectQueryExpressionAnalyzer::makeTableJoin(
...
+            joined_plan.reset();
     }
     else
-    {
-        const ColumnsWithTypeAndName & right_sample_columns = subquery_for_join.sample_block.getColumnsWithTypeAndName();
-        bool need_convert = syntax->analyzed_join->applyJoinKeyConvert(left_sample_columns, right_sample_columns);
-        if (need_convert)
-            subquery_for_join.addJoinActions(std::make_shared<ExpressionActions>(syntax->analyzed_join->rightConvertingActions()));
-    }
+        syntax->analyzed_join->applyJoinKeyConvert(left_sample_columns, {});
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Interpreters/ExpressionAnalyzer.cpp:878]: Yes, that's what was interesting for me also :)
After refactoring, the code became like this. Let's try to remove it at all and see what will be broken.

File: src/Processors/Transforms/JoiningTransform.h (hunk 5 above)
Comment (vdimir) [reply src/Processors/Transforms/JoiningTransform.h:17]: It's renames to `FillingRightJoinSideTransform`

File: src/Processors/QueryPipeline.cpp [hunk 7]
```diff
@@ -255,6 +262,95 @@ QueryPipeline QueryPipeline::unitePipelines(
...
+
+    /// This counter is needed for every Joining except totals, to decide which Joining will generate non joined rows.
+    auto finish_counter = std::make_shared<JoiningTransform::FinishCounter>(num_streams);
+
+    auto lit = left->pipe.output_ports.begin();
+    auto rit = right->pipe.output_ports.begin();
+
+    for (size_t i = 0; i < num_streams; ++i)
+    {
+        auto joining = std::make_shared<JoiningTransform>(left->getHeader(), join, max_block_size, false, default_totals, finish_counter);
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Processors/QueryPipeline.cpp:321]: Yes, exactly

File: src/Processors/QueryPlan/JoinStep.cpp (hunk 4 above)
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/JoinStep.cpp:29]: `updatePipeline` is implemented via `transformPipeline` for `ITransformingStep`.

File: src/Processors/Transforms/JoiningTransform.h (hunk 5 above)
Comment (KochetovNicolai) [reply src/Processors/Transforms/JoiningTransform.h]: fixed

**Comment by alexey-milovidov** [comment]
@vdimir @KochetovNicolai This refactoring introduced performance regression:
https://github.com/ClickHouse/ClickHouse/issues/24075

**Pull request merged**